pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -0,0 +1,434 @@
+                    .into_iter()
+                    .map(|layer_infos| (tl.timeline_id, layer_infos)),
+            );
+            if cancel.is_cancelled() {
+                return Ok(IterationOutcome::Cancelled);
+            }
+        }
+    }
+
+    // sort layers by last activity
+    for (_, layers) in &mut data {
+        layers.sort_unstable_by_key(|(_, layer_info)| layer_info.last_activity_ts);
+        if cancel.is_cancelled() {
+            return Ok(IterationOutcome::Cancelled);
+        }
+    }
+
+    // carve out the per-tenant reservation and throw everything else into a global list of LRU candidates
+    let mut lru_candidates = Vec::new();
+    for (tenant_id, layers) in &data {
+        let min_resident_size = task_config.tenant_min_resident_size; // XXX allow per-tenant override, including "unlimited"
+        let mut post_evict_resident_size: u64 =
+            layers.iter().map(|(_, layer)| layer.file_size).sum();
+        for (timeline_id, layer) in layers {
+            if cancel.is_cancelled() {
+                return Ok(IterationOutcome::Cancelled);
+            }
+            if post_evict_resident_size <= min_resident_size {
+                continue;
+            }
+            lru_candidates.push((tenant_id, timeline_id, layer));
+            post_evict_resident_size -= layer.file_size;
+        }
+    }
+
+    // do LRU eviction
+    lru_candidates.sort_unstable_by_key(|(_, _, layer)| layer.last_activity_ts);
+    let mut usage_expected = usage_pre.usage.clone();
+    let mut held_tenants = HashMap::new();
+    let mut held_timelines = HashMap::new(); // cache held timelines; XXX limit size?
+    for (tenant_id, timeline_id, layer) in lru_candidates {
+        let tenant = match held_tenants.entry(tenant_id) {
+            Entry::Occupied(e) => e.into_mut(),
+            Entry::Vacant(e) => {
+                let tenant = match tenant::mgr::get_tenant(*tenant_id, false).await {
+                    Ok(tenant) => tenant,
+                    Err(e) => {
+                        warn! {%tenant_id, "failed to get tenant: {:#}", e};
+                        continue;
+                    }
+                };
+                e.insert(tenant)
+            }
+        };
+        let timeline = match held_timelines.entry(timeline_id) {
+            Entry::Occupied(e) => e.into_mut(),
+            Entry::Vacant(e) => {
+                let timeline = match tenant.get_timeline(*timeline_id, false) {
+                    Ok(timeline) => timeline,
+                    Err(e) => {
+                        warn! {%tenant_id, %timeline_id, "failed to get timeline: {:#}", e};
+                        continue;
+                    }
+                };
+                e.insert(timeline)
+            }
+        };
+        // XXX batching? but, if the batch fails due to one layer, we should retry the remaining ones
+        match timeline.evict_layer(&layer.file_name.file_name()).await {
```

Comment:
I think we could hold on to the `Arc<dyn PersistentLayer>` here. That saves us the layer file name search in `fn evict_layer`. If the `Arc<dyn PersistentLayer>` is no longer in the layer map, eviction will fail with the same error we're handling already.

Reply:
Yes, this would be very much sensible.

Reply:
This has been now done with the changes to use batching yesterday, resolving.